src/System.Management.Automation/utils/Telemetry.cs

```diff
@@ -0,0 +1,403 @@
+        private const string _anonymous = "anonymous";
+
+        // the telemetry failure string
+        private const string _telemetryFailure = "TELEMETRY_FAILURE";
+
+        // Telemetry client to be reused when we start sending more telemetry
+        private static TelemetryClient s_telemetryClient { get; set; }
+
+        // the unique identifier for the user, when we start we
+        private static string s_uniqueUserIdentifier { get; set; }
+
+        // the session identifier
+        private static string s_sessionId { get; set; }
+
+        /// Use a hashset for quick lookups.
+        /// We send telemetry only a known set of modules.
+        /// If it's not in the list (initialized in the static constructor), then we report anonymous.
+        private static HashSet<string> s_knownModules;
+
+        /// <summary>Gets a value indicating whether telemetry can be sent.</summary>
+        public static bool CanSendTelemetry { get; private set; }
+
+        /// <summary>
+        /// Initializes static members of the <see cref="ApplicationInsightsTelemetry"/> class.
+        /// Static constructor determines whether telemetry is to be sent, and then
+        /// sets the telemetry key and set the telemetry delivery mode.
+        /// Creates the session ID and initializes the HashSet of known module names.
+        /// Gets or constructs the unique identifier.
+        /// </summary>
+        static ApplicationInsightsTelemetry()
+        {
+            // If we can't send telemetry, there's no reason to do any of this
+            CanSendTelemetry = !GetEnvironmentVariableAsBool(name: _telemetryOptoutEnvVar, defaultValue: false);
+            if (CanSendTelemetry)
+            {
+                s_telemetryClient = new TelemetryClient();
+                TelemetryConfiguration.Active.InstrumentationKey = _psCoreTelemetryKey;
+                // Set this to true to reduce latency during development
+                TelemetryConfiguration.Active.TelemetryChannel.DeveloperMode = false;
+                s_sessionId = Guid.NewGuid().ToString();
+
+                // use a hashset when looking for module names, it should be quicker than a string comparison
+                s_knownModules = new HashSet<string>(
+                    new string []
+                    {
+                        "Microsoft.PowerShell.Archive",
+                        "Microsoft.PowerShell.Host",
+                        "Microsoft.PowerShell.Management",
+                        "Microsoft.PowerShell.Security",
+                        "Microsoft.PowerShell.Utility",
+                        "PackageManagement",
+                        "Pester",
+                        "PowerShellGet",
+                        "PSDesiredStateConfiguration",
+                        "PSReadLine",
+                        "ThreadJob",
+                    },
+                    StringComparer.OrdinalIgnoreCase
+                    );
```

Comment:
I recommend to use the collection initializer syntax, as in
```c#
s_knownModules = new HashSet<string>(StringComparer.OrdinalIgnoreCase) {
    "Microsoft.PowerShell.Archive",
    "Microsoft.PowerShell.Host",
    ...
};
```

Reply:
fixed